### ISSUE TITLE
rate/burst buggy

### ISSUE THREAD

**Issue by Wazarr94** [haxball/haxball-issues#855]
On game startup, I use the command /kick_rarelimit 0 180 3 (rate 180 burst 3).

Not thanks to the rather poor documentation, I figured you start with the number of the burst (poor name, should be refering to the stock of shots available or something if I believe what it means) that you can spend as you begin the game and then you need to wait.

Here's my plan. I shoot 3 times before the 3 second mark, let's say I have a special map and I can shoot instantly after every shot. I should be hitting the ball 3 times, and then run out of shot.
If I understand this rate/burst thing, I have to wait 3 seconds after my first shot to be able to shoot again, or I can wait 6 seconds so I can shoot twice, or at last, I can wait 9 seconds so I can shoot 3 times in a row.

If I diminish rate to 120, I would have to wait 120 frames (or 2 seconds) to have a new shot, and if I change burst to 5, at startup I can do 5 kicks at the start and then I can load up to 5 shots.

I feel like this is what should be happening, the documentation is a bit too vague, but I can understand since it is hard to explain.

The problem I find is at the start, using my first command (0 180 3), I have a different number of shots everytime I stop and start the game.

There might be something with burst not being reset at the start of each game, I'd bet a dollar on that.

**Comment by basro** [comment]
The rate limit is not reset when you start a new game, I don't consider this a bug.

It might be an enhancement to reset it on kickoffs though, I'll consider it for future updates.

**Comment by Wazarr94** [comment]
Yes, I'm not talking about the rate, but only the number of shots stored. 
If you start and stop multiple times, you can end up with 9 shots in a row possible.

I'm gonna try and record it if possible

**Comment by basro** [comment]
That is what I meant, the stored kicks is not reset and I don't consider it a bug.

burst 0 makes it work like min, so basically the larger of the two will dictate how often you can kick.

**Comment by Wazarr94** [comment]
Doesn't burst 1 makes it the same ? Hence the question.

Update: switched to 0 240 6, started the game, ended the game, started another one and ended it after, switched to 0 120 1 and had a big amount of shots available for some reason

**Comment by Wazarr94** [comment]
OK, after more trys, I understand the answer.

The only problem with it seems that after restarting a game, you do not have your burst reset to the max.

Example of the problem (I tested it but I cannot make a gif from it, my Chrome is refusing and i cba to upload it as a rec) that you can reproduce at home:

1) Create a room
2) Load a map that enables you to shoot at multiple times quickly without much distance to the ball
3) Enter the command /kick_ratelimit 0 240 2. When fully loaded up (you gotta wait 6 seconds), you get 1 normal shot + 2 extra shots (burst value of 2) 
4) Start the game

If you rush the ball, you only get 1 shot (you don't start fully loaded up), if you wait 6 seconds, you'll get 3 shots.
If you stop the game right after doing all your shots, if you start the game, you will not have the possibility to shot the ball for some time.
The correct behavior imo should be that you get all shots at the reset of position, this way this behavior cannot happen.
Also, if you change kick rate limit in a middle of a game, there are some weird stuff happening.

But as long as no one figures it out, I think this is a minor bug yes

**Comment by Wazarr94** [comment]
Final post (even minor problem), the replay doesn't show the change of kick_ratelimit, even though the game shows it.

Thanks

**Comment by basro** [comment]
Ok yes this is a bug. Changing the kick ratelimit of a room doesn't reset some internal counters which means players can surpass the rate limit for the first few kicks.

This bug requires a backwards incompatible update to fix and it is low priority because the only way to "exploit" it is to first set a high kick rate limit and then lower it.

I will fix it the next time I make a backwards incompatible update.

**Comment by Wazarr94** [comment]
No problem, I was just playing around and saw weird stuff happening, just wanted to mention it.
